diff --git a/sys/kern/kern_jail.c b/sys/kern/kern_jail.c
index 9784f3bfc23b..7a49d92334b5 100644
--- a/sys/kern/kern_jail.c
+++ b/sys/kern/kern_jail.c
@@ -215,2 +215,3 @@ static struct bool_flags pr_flag_allow[NBBY * NBPW] = {
 	{"allow.suser", "allow.nosuser", PR_ALLOW_SUSER},
+	{"allow.kmem", "allow.nokmem", PR_ALLOW_KMEM_ACCESS},
 };
@@ -3710,2 +3711,23 @@ prison_priv_check(struct ucred *cred, int priv)
 
+		/*
+		 * Allow access to /dev/io in a jail if the non-jailed admin
+		 * requests this and if /dev/io exists in the jail. This
+		 * allows Xorg to probe a card.
+		 */
+	case PRIV_IO:
+		if (cred->cr_prison->pr_allow & PR_ALLOW_KMEM_ACCESS)
+			return (0);
+		else
+			return (EPERM);
+
+		/*
+		 * Allow low level access to KMEM-like devices (e.g. to
+		 * allow Xorg to use DRI).
+		 */
+	case PRIV_KMEM_WRITE:
+		if (cred->cr_prison->pr_allow & PR_ALLOW_KMEM_ACCESS)
+			return (0);
+		else
+			return (EPERM);
+
 		/*
@@ -3991,2 +4013,6 @@ SYSCTL_PROC(_security_jail, OID_AUTO, mount_allowed,
     "Processes in jail can mount/unmount jail-friendly file systems (deprecated)");
+SYSCTL_PROC(_security_jail, OID_AUTO, kmem_allowed,
+    CTLTYPE_INT | CTLFLAG_RW | CTLFLAG_MPSAFE,
+    NULL, PR_ALLOW_KMEM_ACCESS, sysctl_jail_default_allow, "I",
+    "Processes in jail can access kmem");
 
@@ -4147,2 +4173,4 @@ SYSCTL_JAIL_PARAM(_allow, suser, CTLTYPE_INT | CTLFLAG_RW,
     "B", "Processes in jail with uid 0 have privilege");
+SYSCTL_JAIL_PARAM(_allow, kmem, CTLTYPE_INT | CTLFLAG_RW,
+    "B", "Jail may access kmem-like devices (io, dri) if they exist");
 
diff --git a/usr.sbin/jail/jail.8 b/usr.sbin/jail/jail.8
index 61fe91cf973e..28d01daa4a16 100644
--- a/usr.sbin/jail/jail.8
+++ b/usr.sbin/jail/jail.8
@@ -587,2 +587,13 @@ and resource limits.
 The jail root may bind to ports lower than 1024.
+.It Va allow.kmem
+Jailed processes may access
+.Pa /dev/kmem
+and similar devices (e.g. io, dri) if they have sufficient permission
+(via the usual file permissions).
+Note that the device files must exist within the jail for this parameter
+to be of any use;
+the default devfs ruleset for jails does not include any such devices.
+Giving a jail access to kernel memory obviates much of the security that
+jails offer, but can still be useful for other purposes.
+For example, this would allow the Xorg server to run inside a jail.
 .It Va allow.unprivileged_proc_debug
